### ISSUE TITLE
Undefined symbol: __tls_base 

### ISSUE THREAD

**Issue by davlhd** [emscripten-core/emscripten#12489]
I am trying to compile OpenCV 3.1.0 and use it in a C++ project. I am using 2.0.5 and I am getting this error when linking:

undefined symbol: __tls_base (referenced by top-level compiled C/C++ code)

I tried switching optimizations but the error persists.

Do you know what could cause that?

( I am using docker and ubuntu:20.04)

**Comment by davlhd** [comment]
Downgrading to 2.0.0 solved the problem

**Comment by sbc100** [comment]
Are you trying to build a multi-threaded program?  (i.e. are you liking with `-s USE_PTHREADS=1`)?

When you upgrade from 2.0.0 to 2.0.5 are you re-building all object files and library files?  Does it fix the issue if you do?  (Not that you should have to do this in general but I'm just curious if that solves the issue... as that information might help us in debugging).

**Comment by davlhd** [comment]
I am not using USE_PTHREADS and yes, I rebuild everything every time I change the compiler.

**Comment by sbc100** [comment]
You must be linking some object built with `-pthread` somehow.

Can you add `-s LLD_REPORT_UNDEFINED_SYBMOLS` so that the linker might tell you which file contains this undefined symbol?

**Comment by anthann** [comment]
I'm using sdk of version emscripten-master-64bit to build latest opencv.js(4.4.0), and I got the same error. 
After adding "-s LLD_REPORT_UNDEFINED", the linker shows as blow:

`wasm-ld: error: entry symbol not defined (pass --no-entry to suppress): main
em++: error: '/Users/abc/workspace/lib/emsdk/llvm/git/build_master_64/bin/wasm-ld -o /var/folders/sz/v5tr178x1fj40x70bxwr90bc0000gn/T/emscripten_temp_65jeh4md/opencv_js.wasm --gc-sections CMakeFiles/opencv_js.dir/bindings.cpp.o -L/Users/abc/workspace/lib/emsdk/emscripten/master/system/local/lib -L/Users/abc/workspace/lib/emsdk/emscripten/master/system/lib ../../lib/libopencv_core.a -L/Users/abc/workspace/lib/emsdk/emscripten/master/cache/wasm ../../lib/libopencv_flann.a ../../lib/libopencv_imgproc.a ../../lib/libopencv_photo.a ../../lib/libopencv_dnn.a ../../lib/libopencv_features2d.a ../../lib/libopencv_calib3d.a ../../lib/libopencv_objdetect.a ../../lib/libopencv_video.a ../../3rdparty/lib/liblibprotobuf.a ../../lib/libopencv_calib3d.a ../../lib/libopencv_features2d.a ../../lib/libopencv_flann.a ../../lib/libopencv_imgproc.a ../../lib/libopencv_core.a ../../3rdparty/lib/libzlib.a --whole-archive /Users/abc/workspace/lib/emsdk/emscripten/master/cache/wasm/libembind-rtti.a --no-whole-archive /Users/abc/workspace/lib/emsdk/emscripten/master/cache/wasm/libc.a /Users/abc/workspace/lib/emsdk/emscripten/master/cache/wasm/libcompiler_rt.a /Users/abc/workspace/lib/emsdk/emscripten/master/cache/wasm/libc++-noexcept.a /Users/abc/workspace/lib/emsdk/emscripten/master/cache/wasm/libc++abi-noexcept.a /Users/abc/workspace/lib/emsdk/emscripten/master/cache/wasm/libdlmalloc.a /Users/abc/workspace/lib/emsdk/emscripten/master/cache/wasm/libpthread_stub.a /Users/abc/workspace/lib/emsdk/emscripten/master/cache/wasm/libc_rt_wasm.a /Users/abc/workspace/lib/emsdk/emscripten/master/cache/wasm/libsockets.a -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --allow-undefined-file=/var/folders/sz/v5tr178x1fj40x70bxwr90bc0000gn/T/tmpt5ztidj1.undefined --import-memory --strip-debug --export-table --export main --export __cxa_demangle --export stackSave --export stackRestore --export stackAlloc --export __data_end --export __wasm_call_ctors --export __errno_location --export malloc --export free --export setThrew --export emscripten_main_thread_process_queued_calls -z stack-size=5242880 --initial-memory=134217728 --entry=main --max-memory=1073741824 --global-base=1024' failed (1)
make[3]: *** [bin/opencv_js.js] Error 1
make[2]: *** [modules/js/CMakeFiles/opencv_js.dir/all] Error 2
make[1]: *** [modules/js/CMakeFiles/opencv.js.dir/rule] Error 2
make: *** [opencv.js] Error 2
Traceback (most recent call last):
  File "./platforms/js/build_js.py", line 277, in <module>
    builder.build_opencvjs()
  File "./platforms/js/build_js.py", line 195, in build_opencvjs
    execute(["make", "-j", str(multiprocessing.cpu_count()), "opencv.js"])
  File "./platforms/js/build_js.py", line 23, in execute
    raise Fail("Child returned: %s" % retcode)
__main__.Fail: Child returned: 2`

**Comment by anthann** [comment]
By reproduce this problem, you could just test as below:

1. use latest emsdk
2. git clone https://github.com/opencv/opencv.git
3. cd opencv/
4. python ./platforms/js/build_js.py  --build_wasm --emscripten_dir PATH_TO_UPSTREAM  build_wasm

**Comment by tlively** [comment]
wasm-ld emits `__tls_base` as a Wasm global variable, not as a normal defined-data symbol. I'm not surprised referencing it from source code doesn't work. @sbc100 do we have any way to reference Wasm globals as normal data symbols in C++?

**Comment by sbc100** [comment]
No you can't reference that symbols from C++ ... but it is still a symbol, just like `__stack_pointer` and the compiler generates references to it with it compiles with threads enabled right?

The error message `referenced by top-level compiled C/C++ code` is a bit misleading.. what it means is that the undefined symbol is in the output of the linker.. as opposed to the JS library code.

**Comment by Isaac-the-Man** [comment]
Got the same error when building OpenCV.js 4.4.0. I'm using cmake 3.16 and latest emsdk. 
```
python ./platforms/js/build_js.py build_js --emscripten_dir=EMSCRIPTEN_DIR
```

**Comment by tlively** [comment]
With this sample program:

```
#include <stdio.h>

extern void* __tls_base;

int main() {
  printf("%p\n", __tls_base);
}
```

Compiling *without* `-pthread` worked fine, but compiling with `-pthread` gave me

```
wasm-ld: error: symbol type mismatch: __tls_base
>>> defined as WASM_SYMBOL_TYPE_GLOBAL in <internal>
>>> defined as WASM_SYMBOL_TYPE_DATA in /tmp/emscripten_temp_swv9sk_n/stack-local_0.o
```

**Comment by tlively** [comment]
Also, when I followed the instructions from this [comment](https://github.com/emscripten-core/emscripten/issues/12489#issuecomment-707627525), I got the following error:

```
-- OpenCV: system-specific configuration file is not found: 'Emscripten'
-- Detected processor: x86
-- Looking for ccache - not found
CMake Error at cmake/OpenCVCompilerOptimizations.cmake:593 (message):
  Compiler doesn't support baseline optimization flags:
Call Stack (most recent call first):
  cmake/OpenCVCompilerOptions.cmake:296 (ocv_compiler_optimization_options)
  CMakeLists.txt:610 (include)
```

**Comment by perceptoid** [comment]
The problem seems have been introduced between 2.0.4 and 2.0.5. 

2.0.4 is the most current version, with which I am able to built OpenCV.js 4.4.0. 

2.0.5 through 2.0.7 all give me the mentioned __tls_base linker error.

**Comment by tlively** [comment]
@sbc100 would you be able to look into this more?

**Comment by sbc100** [comment]
Sure I'll take a look.

**Comment by anthann** [comment]
If I comment out [this line](https://github.com/opencv/opencv/blob/d5fd2f0155ffad366f9ac912dfd6d189a7a6a98e/cmake/OpenCVCompilerOptions.cmake#L170) in opencv's cmake, then no error anymore.
Maybe something related to pthread makes things mass?

**Comment by Slimmmo** [comment]
I am also getting this error trying to link an emcc compiled protobuf library to my code. Also not using pthreads (unless protobuf is compiled with them by default).

**Comment by sbc100** [comment]
OK so it looks like at least one of your object files uses atomics and TLS (i.e. was built with threading support).  Specifically `CMakeFiles/libprotobuf.dir/src/google/protobuf/arena.cc.o` referes to `__tls_base`.

We used to error out this case and report:   `'atomics' feature is used, so --shared-memory must be used`

However since https://reviews.llvm.org/D79530 we no longer error out since atomics are now supported in single threaded builds too.   However, I guess TLS is not.

Sort term solution:  Figure out why `arena.cc.o` is being built with atomics enabled and stop doing that.

Long term solution: Figure out what the linker is supposed to do in this case.

**Comment by jwtowner** [comment]
I'm getting this exact same problem again in 2.0.30. The code is using atomics, but no thread_local/_Thread_local storage, while exhibiting the same linker error.

**Comment by sbc100** [comment]
@jwtowner can you explain a little bit more about use case?   Can you perhaps show the entire like command along with its full output (including the full error message).

Are you linking with `-s USE_PTHREADS` (or `-pthreads`) ?  Are you compiling with `-s USE_PTHREADS` (or `-pthreads`) ?

**Comment by jwtowner** [comment]
> Are you linking with `-s USE_PTHREADS` (or `-pthreads`) ? Are you compiling with `-s USE_PTHREADS` (or `-pthreads`) ?

Yep, I'm not using `-s USE_PTHREADS` or `-pthreads`, this is for a single-threaded build. With `-s USE_PTHREADS` for our multi-threaded build it links fine. It also works fine without -O3 and -flto I believe. I'm going to try downgrading to 2.0.27 and see if it the single-threaded build still links there (last known version that still worked for us). Unfortunately, on 2.0.29 we began hitting a completely different internal linker error that affects both our multi-threaded and single-threaded builds that seems to be fixed in 2.0.30, but 2.0.30 appears to have this particular regression.

Our full command line looks something like this:

```em++ -o <OutputPath>.js -Wl,--strip-debug -Wl,-z,relro -Wl,-z,noexecstack -s ASYNCIFY=1 -s FORCE_FILESYSTEM=1 -s MAX_WEBGL_VERSION=2 -lfs.js -lmemfs.js -lidbfs.js -lsockfs.js -s ENVIRONMENT=web -O3 -flto -Wl,--lto-O3 -s ASSERTIONS=0 -s STACK_OVERFLOW_CHECK=1 -s TOTAL_MEMORY=402653184 -s MODULARIZE=1 -s EXPORT_NAME="createModule" -s USE_CLOSURE_COMPILER=1 -s IGNORE_CLOSURE_COMPILER_ERRORS=1 --closure 1 -Wl,-error-limit=0 --pre-js <PreloadScript>.js <Library & Object Dependencies>```

Thanks for the quick reply.

**Comment by sbc100** [comment]
Hmm...  I'm having trouble seeing how references to `__tls_base` can exist in lto.tmp.   I thought that all TLS should be lowered away.  Investigating.

**Comment by jwtowner** [comment]
Compiling and linking everything with -flto=thin avoids the `undefined symbol: __tls_base` error, but getting the same linker warnings as described in https://github.com/emscripten-core/emscripten/issues/10831 and a memory access out of bounds exception immediately after running. Going to try disabling LTO altogether and see what happens.

```
Uncaught RuntimeError: memory access out of bounds
    at showdown-unicore.wasm:0x8676cc
    at b.<computed> (showdown-unicore.js:236)
    at h._main (showdown-unicore.js:339)
    at b (showdown-unicore.js:352)
    at showdown-unicore.js:353
```

**Comment by sbc100** [comment]
I managed to reproduce the issue.  Working on a wasm-ld fix now.

**Comment by jwtowner** [comment]
Fyi, my current work around is just to disable LTO for the Unicore builds.

**Comment by sbc100** [comment]
The llvm change is here: https://reviews.llvm.org/D111171.  I'm still working on the test but the fix is good AFAICT.

**Comment by sbc100** [comment]
I can't seem to reproduce the issue.   The problem is that in all the tests I've managed to create all the TLS variables get lowered away by https://github.com/llvm/llvm-project/blob/f92961d238efdfeccce27f9bb7b0a6629c376d4a/llvm/lib/Target/WebAssembly/WebAssemblyTargetMachine.cpp#L295.. and therefore `WebAssemblyTargetLowering::LowerGlobalTLSAddress` is never called.. and AFAICT that is the only reasonable place that those references to `__tls_base` could be created during LTO.

@jwtowner , is there any way you could build llvm yourself to try to figure out what is going on?     Or maybe you could upload zip file containing all the object file inputs needed to create this undefined symbol failure.

In short, I can see how it can happen if the reference to `__tls_base` is created during LTO, but I can't see how that could ever happen since I can't see how `LowerGlobalTLSAddress` is ever called.

**Comment by sbc100** [comment]
OK I figured it out and landed the fix: https://reviews.llvm.org/D111171
